arwen/contexts/instanceTracker.go

```diff
@@ -56,54 +62,56 @@ func NewInstanceTracker() (*instanceTracker, error) {
 	return tracker, nil
 }
 
+// InitState initializes the internal instanceTracker state
 func (tracker *instanceTracker) InitState() {
+	tracker.instance = nil
 	tracker.codeHash = make([]byte, 0)
 }
 
+// PushState pushes the active instance and codeHash on the state stacks
 func (tracker *instanceTracker) PushState() {
 	tracker.instanceStack = append(tracker.instanceStack, tracker.instance)
 	tracker.codeHashStack = append(tracker.codeHashStack, tracker.codeHash)
-	logRuntime.Trace("pushing instance", "id", tracker.instance.Id(), "codeHash", tracker.codeHash)
+	logTracker.Trace("pushing instance", "id", tracker.instance.ID(), "codeHash", tracker.codeHash)
 }
 
+// PopSetActiveState pops the instance and codeHash from the state stacks and sets them as active
 func (tracker *instanceTracker) PopSetActiveState() {
 	instanceStackLen := len(tracker.instanceStack)
 	if instanceStackLen == 0 {
 		return
 	}
 
-	prevInstance := tracker.instanceStack[instanceStackLen-1]
-	tracker.instanceStack = tracker.instanceStack[:instanceStackLen-1]
+	activeInstance := tracker.instance
+	activeCodeHash := tracker.codeHash
+	stackedPrevInstance := tracker.instanceStack[instanceStackLen-1]
 
-	if prevInstance == tracker.instance {
-		// The current Wasmer instance was previously pushed on the instance stack,
-		// but a new Wasmer instance has not been created in the meantime. This
-		// means that the instance at the top of the stack is the same as the
-		// current instance, so it cannot be cleaned, because the execution will
-		// resume on it. Popping will therefore only remove the top of the stack,
-		// without cleaning anything.
-		return
-	}
-
-	if !check.IfNil(tracker.instance) {
-		onStack := tracker.IsCodeHashOnTheStack(tracker.codeHash)
-		coldOnlyEnabled := !WarmInstancesEnabled
-		if onStack || coldOnlyEnabled {
-			if tracker.instance.Clean() {
-				tracker.updateNumRunningInstances(-1)
-			}
-		}
+	onStack := tracker.IsCodeHashOnTheStack(activeCodeHash)
+	activeInstanceIsTopOfStack := stackedPrevInstance == activeInstance
+	cold := !WarmInstancesEnabled
```

Comment:
why does this global exists ? we should not have something like this. Warm instance is always enabled.

Reply:
This is the emergency kill-switch for warm instances. Without it, warm instances cannot be disabled easily if another memory leak is discovered (or other problems).

But yes, this constant is always set to `true`.

Reply:
It was useful for debugging purposes